File: go/examples/simple_workflow.go
package main

import (
	"fmt"
	"os"
	"time"

	"workflowengine/engine"
)

func greetingWorkflow(ctx *engine.WorkflowContext, input map[string]interface{}) (interface{}, error) {
	version, err := ctx.GetVersion("greeting-change", 1, 2)
	if err != nil {
		return nil, err
	}
	name := input["name"].(string)
	greeting, err := ctx.ExecuteActivity("compose_greeting", map[string]interface{}{"name": name}, nil)
	if err != nil {
		return nil, err
	}
	_ = ctx.Sleep(1)
	approval, err := ctx.WaitSignal("approval")
	if err != nil {
		return nil, err
	}
	return map[string]interface{}{
		"greeting":    greeting,
		"approved_by": approval["user"],
		"version":     version,
	}, nil
}

func composeGreeting(ctx *engine.ActivityContext, input map[string]interface{}) (interface{}, error) {
	_ = ctx.Heartbeat(map[string]interface{}{"stage": "starting"})
	return "Hello, " + input["name"].(string) + "!", nil
}

func main() {
	if _, err := os.Stat("example.db"); err == nil {
		_ = os.Remove("example.db")
	}
	config := engine.DefaultConfig()
	config.DBPath = "example.db"
	config.TaskQueueName = "default"
	config.WorkerID = "go-worker"

	store, err := engine.NewSQLiteStore(config.DBPath)
	if err != nil {
		panic(err)
	}
	metrics := engine.NewMetricsRegistry()
	queue := engine.NewDBTaskQueue(store.DB(), metrics)
	engineSvc := engine.NewWorkflowEngine(store, queue, config, metrics)
	worker := engine.NewWorker(store, queue, config, metrics)
	scheduler := engine.NewTimerScheduler(store, queue, config, metrics)

	worker.RegisterWorkflow("GreetingWorkflow", greetingWorkflow)
	worker.RegisterActivity("compose_greeting", composeGreeting)

	worker.Start()
	scheduler.Start()

	runID, err := engineSvc.StartWorkflow("greeting-workflow-1", "GreetingWorkflow", map[string]interface{}{"name": "Temporal"}, "")
	if err != nil {
		panic(err)
	}
	time.Sleep(2 * time.Second)
	_ = engineSvc.SignalWorkflow("greeting-workflow-1", runID, "approval", map[string]interface{}{"user": "admin@example.com"})

	for {
		exec, _ := store.GetWorkflowExecution("greeting-workflow-1", runID)
		if exec != nil && exec.State == engine.WorkflowStateCompleted {
			fmt.Println("Workflow result:", exec.Result.String)
			break
		}
		time.Sleep(500 * time.Millisecond)
	}
	fmt.Println("Metrics:", metrics.Snapshot())
}

File: python/examples/simple_workflow.py
import os
import time

from engine import (
    Config,
    DBTaskQueue,
    MetricsRegistry,
    SQLiteStore,
    TimerScheduler,
    WorkflowEngine,
    Worker,
    configure_logging,
)


def greeting_workflow(ctx, input_payload):
    # Versioning API ensures backward compatibility across workflow code changes.
    version = yield ctx.get_version("greeting-change", 1, 2)
    name = input_payload.get("name", "world")
    greeting = yield ctx.execute_activity(
        "compose_greeting",
        {"name": name},
        schedule_to_close_timeout_seconds=30,
        start_to_close_timeout_seconds=10,
        heartbeat_timeout_seconds=5,
    )
    yield ctx.sleep(1)
    approval = yield ctx.wait_signal("approval")
    return {
        "greeting": greeting,
        "approved_by": approval.get("user"),
        "version": version,
    }


def compose_greeting(activity_ctx, input_payload):
    activity_ctx.heartbeat({"stage": "starting"})
    return f"Hello, {input_payload['name']}!"


def run_example():
    configure_logging("INFO")
    if os.path.exists("example.db"):
        os.remove("example.db")
    config = Config(db_path="example.db", task_queue_name="default", worker_id="python-worker")
    metrics = MetricsRegistry()
    store = SQLiteStore(config.db_path)
    queue = DBTaskQueue(store.connection, metrics=metrics)
    engine = WorkflowEngine(store, queue, config, metrics=metrics)
    worker = Worker(store, queue, config, metrics=metrics)
    worker.register_workflow("GreetingWorkflow", greeting_workflow)
    worker.register_activity("compose_greeting", compose_greeting)
    scheduler = TimerScheduler(store, queue, config, metrics=metrics)

    scheduler.start()
    worker.start()

    run_id = engine.start_workflow(
        workflow_id="greeting-workflow-1",
        workflow_type="GreetingWorkflow",
        input_payload={"name": "Temporal"},
    )

    time.sleep(2)
    engine.signal_workflow(
        workflow_id="greeting-workflow-1",
        run_id=run_id,
        signal_name="approval",
        payload={"user": "admin@example.com"},
    )

    while True:
        execution = store.get_workflow_execution("greeting-workflow-1", run_id)
        if execution and execution["state"] == "COMPLETED":
            print("Workflow result:", execution["result"])
            break
        time.sleep(0.5)

    worker.stop()
    scheduler.stop()
    print("Metrics:", metrics.snapshot())


if __name__ == "__main__":
    run_example()

